core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -495,8 +495,9 @@ assert equalType(
     for (MutableRel targetDescendant : targetDescendants) {
       MutableRel queryDescendant = map.get(targetDescendant);
       if (queryDescendant != null) {
-        assert queryDescendant.rowType.equals(targetDescendant.rowType);
-        equivalents.put(queryDescendant, targetDescendant);
```

Comment:
I made this change, because rel like https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/rel/mutable/MutableAggregate.java#L68 can will equal with each other but have different row type.

Reply:
We can add a test for such exception.

Reply:
Sorry, @jinxing64 , I'm not familiar with this part of the code. So, you're saying that it's ok for the 'else' block here to be a no-op (and not for example, throwing an exception?). 
Maybe a comment line will be helpful to explain the reason behind.

Reply:
@DonnyZone  @rubenada 
THX a lot for review this.
Sure I will add a comment line soon for this.

> We can add a test for this exception.
hmm... Actually what I'm doing is to release the assert and avoid the exception, so maybe it's hard to test the exception catching.

Reply:
@jinxing64 , maybe I'm missing something, but according to your comment we can have two MutableAggregate which are equal but have a different row-type? I'm curious to know how we arrive to this situation, since the equals/hashCode and deriveRowType methods are based on the same fields (groupSet, groupSets, aggCalls, input)

Reply:
> 
> 
> @DonnyZone @rubenada
> THX a lot for review this.
> Sure I will add a comment line soon for this.
> 
> > We can add a test for this exception.
> 
> hmm... Actually what I'm doing is to release the assert and avoid the exception, so maybe it's hard to test the exception catching.

I have also noticed this exception, you can use a similar example as that in [CALCITE-3113](https://issues.apache.org/jira/browse/CALCITE-3113).

Reply:
Thanks for the clarification, @DonnyZone

Reply:
@DonnyZone 
I added a test 
```
  @Test public void testAggregate1() {
    checkMaterialize(
        "select \"deptno\", count(*) as c, sum(\"empid\") as s from \"emps\" group by \"deptno\"",
        "select count(*) + 1 as c, \"deptno\" from \"emps\" group by \"deptno\"");
  }
```
As I removed the row-type checking, exception is not thrown now. Given existing substitution rules, materialization and query above can match with each other.